pkg/vulndump/write.go

```diff
@@ -58,3 +62,166 @@ func WriteOSVulns(outputDir string, vulns []database.Vulnerability) error {
 }
+
+// archive is an adapted implementation of (*Zip).Archive from
+// https://github.com/mholt/archiver/blob/v3.5.1/zip.go#L140
+// under MIT License.
+func archive(sources []string, destination string) error {
+	if !strings.HasSuffix(destination, ".zip") {
+		return errors.Errorf("%s must have a .zip extension", destination)
+	}
+	if fileExists(destination) {
+		return errors.Errorf("file already exists: %s", destination)
+	}
+	dir := filepath.Dir(destination)
+	if !fileExists(dir) {
+		err := os.MkdirAll(dir, 0755)
+		if err != nil {
+			return errors.Wrapf(err, "making directory: %s", dir)
+		}
+	}
+
+	out, err := os.Create(destination)
+	if err != nil {
+		return errors.Wrapf(err, "creating %s", destination)
+	}
+	defer utils.IgnoreError(out.Close)
+
+	zipW := zip.NewWriter(out)
+	zipW.RegisterCompressor(zip.Deflate, func(out io.Writer) (io.WriteCloser, error) {
+		return flate.NewWriter(out, flate.BestCompression)
+	})
+	defer utils.IgnoreError(zipW.Close)
+
+	for _, source := range sources {
+		err := writeWalk(zipW, source, destination)
+		if err != nil {
+			return errors.Wrapf(err, "walking %s", source)
+		}
+	}
+
+	return nil
+}
+
+// fileExists is an adapted implementation of fileExists from
+// https://github.com/mholt/archiver/blob/v3.5.1/archiver.go#L279
+// under MIT License.
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return !errors.Is(err, fs.ErrNotExist)
+}
+
+// fileInfo is an adapted implementation of FileInfo from
+// https://github.com/mholt/archiver/blob/v3.5.1/archiver.go#L125
+// under MIT license.
+type fileInfo struct {
+	os.FileInfo
+	customName string
+}
+
+func (f fileInfo) Name() string {
+	if f.customName != "" {
+		return f.customName
+	}
+	return f.FileInfo.Name()
+}
+
+// writeWalk is an adapted implementation of (*Zip).writeWalk from
+// https://github.com/mholt/archiver/blob/v3.5.1/zip.go#L300
+// under MIT License.
+func writeWalk(zipW *zip.Writer, source, destination string) error {
+	sourceInfo, err := os.Stat(source)
+	if err != nil {
+		return errors.Wrapf(err, "stat: %s", source)
+	}
+	destAbs, err := filepath.Abs(destination)
+	if err != nil {
+		return errors.Wrapf(err, "getting absolute path of destination %s: %s", destination, source)
+	}
+
+	return filepath.Walk(source, func(fpath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return errors.Wrapf(err, "traversing %s", fpath)
+		}
+		if info == nil {
+			return errors.Errorf("%s: no file info", fpath)
+		}
+
+		fpathAbs, err := filepath.Abs(fpath)
+		if err != nil {
+			return errors.Wrapf(err, "%s: getting absolute path", fpath)
+		}
+		if fsutil.Within(fpathAbs, destAbs) {
+			return nil
+		}
+
+		// build the name to be used within the archive
+		nameInArchive, err := makeNameInArchive(sourceInfo, source, "", fpath)
+		if err != nil {
+			return err
+		}
+		finfo := fileInfo{
+			FileInfo:   info,
+			customName: nameInArchive,
+		}
+
+		var file io.ReadCloser
+		if finfo.Mode().IsRegular() {
```

Comment:
In the original source this was `info.Mode().IsRegular()`, curious why the change to `finfo`?